src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -198,47 +238,34 @@ let pp_fun_def ppf Program.({fdrt; fdname; fdargs; fdbody; _})
     (if is_dist || is_lp then ["bool propto__"] else [])
     @ List.(map ~f:typename (argtypetemplates @ extra_templates))
   in
-  let pp_sig ppf name =
-    pp_template_decorator ppf templates ;
-    pp_returntype ppf fdargs fdrt ;
-    let arg_strs =
-      args @ mk_extra_args extra_templates extra @ ["std::ostream* pstream__"]
-    in
-    pf ppf "%s(@[<hov>%a@]) " name (list ~sep:comma string) arg_strs
-  in
-  let pp_sig_rs ppf name =
-    if is_dist then pp_template_decorator ppf (List.tl_exn templates)
+  let pp_sig ppf (name, args, is_reduce_sum_fn, is_variadic_ode_fn) =
+    if (is_reduce_sum_fn || is_variadic_ode_fn) && is_dist then
+      pp_template_decorator ppf (List.tl_exn templates)
     else pp_template_decorator ppf templates ;
     pp_returntype ppf fdargs fdrt ;
-    let first_three, rest = List.split_n args 3 in
-    let arg_strs =
-      first_three
-      @ ["std::ostream* pstream__"]
-      @ rest
-      @ mk_extra_args extra_templates extra
+    let first_args, rest_args =
+      if is_reduce_sum_fn then List.split_n args 3 else List.split_n args 2
     in
```

Comment:
Great to see `pp_sig` and `pp_sig_rs` unified.
I'd change these names to `let args, variadic_args` and make the latter empty if this is neither `is_reduce_sum_fn` nor `is_variadic_ode_fn` as I think an explicit `if is_variadic_ode_fn` makes it clearer where `split_n args 2` comes from.
Also looks like `extra_templates` is non-empty only for `rng` and `lp` functions which are not allowed with variadic functions; the following `let arg_strs =` does not need `if` branching because both options fit the same pattern `first_args @ mk_extra @ ["pstream__"] @ variadic_args`.

Reply:
Yeah, I hated how I did pp_sig and pp_sig_rs then...
Thanks for the suggestions, agree with all.